examples/token/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+macro_rules! abort {
+    ($code:ident, $msg:literal $(, $ex:expr)*) => {
+        fvm_sdk::vm::abort(
+            fvm_shared::error::ExitCode::$code as u32,
+            Some(format!($msg, $($ex,)*).as_str()),
+        )
+    };
+}
+
+/// A macro to CBOR-encode a return value.
+macro_rules! encode {
+    ($ex:expr) => {
+        match RawBytes::serialize($ex) {
+            Ok(ret) => ret,
+            Err(err) => {
+                abort!(ErrSerialization, "failed to encode return value: {:?}", err)
+            }
+        }
+    };
+}
+
+/// The state object of the token actor.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct State {
+    pub name: String,
+    pub symbol: String,
+    #[serde(with = "bigint_ser")]
+    pub max_supply: TokenAmount,
+    // map[ActorID] => TokenAmount
+    pub balances: Cid,
+    // TODO should never use ActorID as allowees, since it's an unstable address
+    //  make sure to check for this
+    //  map[ActorID] => map[Address]TokenAmount
+    // pub allowances: Cid,
+}
+
+/// The actor's WASM entrypoint. It takes the ID of the parameters block,
+/// and returns the ID of the return value block.
+#[no_mangle]
+pub fn invoke(params_id: u32) -> u32 {
+    // First, load the current state root.
+    let root = match sdk::sself::root() {
+        Ok(root) => root,
+        Err(err) => abort!(ErrIllegalState, "failed to get root: {}", err),
+    };
+
+    // Load the actor state from the state tree.
+    let state = match Blockstore.get_cbor::<State>(&root) {
+        Ok(Some(state)) => state,
+        Ok(None) => abort!(ErrIllegalState, "state does not exist"),
+        Err(err) => abort!(ErrIllegalState, "failed to get state: {}", err),
+    };
+
+    // Conduct method dispatch. Handle input parameters and return data.
+    let ret: Option<RawBytes> = match sdk::message::method_number() {
+        1 => Some(name(state)),
+        2 => Some(symbol(state)),
+        3 => Some(max_supply(state)),
+        4 => {
+            let params: TransferParams = match params_cbor(params_id) {
+                Ok(params) => params,
+                Err(err) => abort!(ErrIllegalArgument, "failed to parse params: {:?}", err),
+            };
+            transfer(state, params);
+            None
+        }
+        5 => {
+            let addr: Address = match params_cbor(params_id) {
+                Ok(addr) => addr,
+                Err(err) => abort!(ErrIllegalArgument, "failed to parse address: {:?}", err),
+            };
+            Some(balance_of(state, addr))
+        }
+
+        // TODO this exit code should move to actor space given that dispatch
+        //  is now conducted in-actor.
+        _ => abort!(SysErrInvalidMethod, "unrecognized method"),
+    };
+
+    // Insert the return data block if necessary, and return the correct
+    // block ID.
+    match ret {
+        None => NO_DATA_BLOCK_ID,
+        Some(v) => match sdk::ipld::put_block(DAG_CBOR, v.bytes()) {
```

Comment:
dont disagree. maybe this could be one mode of doing things....... but also check out what steb did up there with the sugar for basically state monad actors.

Reply:
There's already such sugar in the Actors FVM runtime shim: https://github.com/filecoin-project/ref-fvm/blob/master/actors/runtime/src/runtime/fvm.rs#L201. We need to port a version of that over to the FVM SDK, and other alternatives will emerge. This actor deliberately uses the plumbing.